### PR TITLE
Efficient Entry and Easy Backup Changes

### PR DESCRIPTION
**Pull request by julian-george**
# plot-data-table

Adds new data table as an alternative way to enter data and review censuses. Also, huge refactor of plotting sheet, with more precise marker placement, more precise grid lines, and ability to drag and drop existing trees.
 
## Type of Change

- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Documentation update

## Referenced Issue(s)

- [195](https://github.com/dali-lab/inforest/issues/195)
- [194](https://github.com/dali-lab/inforest/issues/194)
- [193](https://github.com/dali-lab/inforest/issues/193)
- [188](https://github.com/dali-lab/inforest/issues/188)
- [187](https://github.com/dali-lab/inforest/issues/187)

## Todo
- Fix bug where marker/tree position gets messed up while initially repositioning

### CODE REVIEW

File: packages/app/components/FailedModal.tsx [hunk 1]
```diff
@@ -59,7 +59,7 @@ const FailedModal: FC = () => {
   } = useAppSelector((state) => state.treeCensusLabels);
   const syncState = useAppSelector((state) => state.sync);
   const getTreeNumFromTree = useCallback(
-    (treeId) => allTrees?.[treeId]?.number || treeId,
+    (treeId) => allTrees?.[treeId]?.tag || treeId,
```
Comment (julian-george) [inline packages/app/components/FailedModal.tsx:62]: Not sure why this still isn't working

File: packages/app/components/PlotDrawer/DataEntryForm/index.tsx [hunk 2]
```diff
@@ -52,7 +61,7 @@ const DataEntryForm: React.FC<DataEntryFormProps & View["props"]> = ({
       <View style={[style, styles.container]}>
         <View style={{ flexDirection: "column" }}>
           <View style={styles.progressRow}>
-            <FormProgress stage={stage} setStage={setStage} />
+            {/* <FormProgress stage={stage} setStage={setStage} /> */}
```
Comment (julian-george) [inline packages/app/components/PlotDrawer/DataEntryForm/index.tsx:64]: We need to examine if this is necessary

File: packages/app/components/PlotDrawer/DataEntryForm/index.tsx [hunk 3]
```diff
@@ -177,24 +188,44 @@ const MetaDataForm: React.FC<MetaFormProps> = ({
...
+          onPress={() =>
+            dispatch(repositioning ? stopRepositioning() : startRepositioning())
+          }
+        >
+          <Ionicons
+            name="move-outline"
+            size={40}
+            color={repositioning ? "#666" : "#000"}
+          />
+        </Pressable>
```
Comment (julian-george) [inline packages/app/components/PlotDrawer/DataEntryForm/index.tsx:228]: is this the best UI for repositioning?

File: packages/app/components/PlottingSheet.tsx [hunk 4]
```diff
@@ -73,18 +86,28 @@ export const PlottingSheet: React.FC<PlottingSheetProps> = ({
...
+
   const [markerPos, setMarkerPos] = useState<{
-    x: number;
-    y: number;
+    // absolute marker coordinates within plot (the same regardless of direction)
+    displayedX: number;
+    displayedY: number;
+    // marker coordinates within current sheet orientation
     adjustedX: number;
     adjustedY: number;
```
Comment (julian-george) [inline packages/app/components/PlottingSheet.tsx:108]: overall, re-examine commenting on this file and make sure everything is as intuitive as possible

File: packages/app/components/PlottingSheet.tsx [hunk 5]
```diff
@@ -97,31 +120,81 @@ export const PlottingSheet: React.FC<PlottingSheetProps> = ({
...
+            0.01 *
+            (sheetSize /
+              Math.sqrt(Math.pow(plot.length, 2) + Math.pow(plot.width, 2))) *
+            FOLIAGE_MAGNIFICATION,
+          MIN_DOT_SIZE
+        ),
+        MAX_DOT_SIZE
+      ),
+    [sheetSize, plot]
+  );
```
Comment (julian-george) [inline packages/app/components/PlottingSheet.tsx:140]: just like the comment says: re-examine this

File: packages/app/components/PlottingSheet.tsx [hunk 6]
```diff
@@ -97,31 +120,81 @@ export const PlottingSheet: React.FC<PlottingSheetProps> = ({
...
+
+  // Rotates the marker's pixel coordinates based on the desired direction
+  const rotateXY = useCallback(
+    (adjustedX: number, adjustedY: number, dir = direction) => {
+      let temp;
+      let transformedX = adjustedX;
+      let transformedY = adjustedY;
+      dir = dir % 4;
```
Comment (julian-george) [inline packages/app/components/PlottingSheet.tsx:148]: maybe change this to account for possible negative directions

File: packages/app/components/PlottingSheet.tsx [hunk 7]
```diff
@@ -73,18 +86,28 @@ export const PlottingSheet: React.FC<PlottingSheetProps> = ({
...
-    y: number;
+    // absolute marker coordinates within plot (the same regardless of direction)
+    displayedX: number;
+    displayedY: number;
+    // marker coordinates within current sheet orientation
     adjustedX: number;
     adjustedY: number;
   }>();
+  // Converts the marker's coordinates in pixels on the sheet to coordinates in meters
   const markerLocToMeters = useCallback(
```
Comment (julian-george) [inline packages/app/components/PlottingSheet.tsx:111]: re-examine, whole "HORIZONTAL" vs "VERTICAL" input seems clunky

File: packages/app/components/fields/FieldController.tsx [hunk 8]
```diff
@@ -32,20 +39,22 @@ export const FieldController: React.FC<FieldControllerProps> = ({
...
   // currValue is the temporary value within the form before it is pushed to the redux store
   const [currValue, setCurrValue] = useState<string>(value);
   // editing
   const [editing, setEditing] = useState(false);
   useEffect(() => {
     const subscription = Keyboard.addListener("keyboardWillHide", () => {
-      setEditing(false);
+      // onConfirm(currValue);
+      // setEditing(false);
     });
```
Comment (julian-george) [inline packages/app/components/fields/FieldController.tsx:52]: re-examine, submitting still feels slow because of this

**Review by julian-george — COMMENTED** [review]
Think about breaking `PlottingSheet` into different files, or at least ordering the callbacks and hooks and stuff better

File: packages/server/src/db/migrations/20220810064741-nullable-species-code.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    try {
+      const firstSpeciesCode = (
+        await queryInterface.sequelize.query(
+          "SELECT code FROM tree_species LIMIT 1",
+          { transaction }
+        )
+      )[0][0].code;
+      await queryInterface.sequelize.query(
+        `UPDATE trees SET "speciesCode" = '${firstSpeciesCode}' WHERE "speciesCode" IS NULL`
+      );
```
Comment (eric-lu-VT) [inline packages/server/src/db/migrations/20220810064741-nullable-species-code.js:40]: What is purpose of `firstSpeciesCode`?

File: packages/server/src/db/migrations/20220726174811-duplicate-census-constraint.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+"use strict";
+
+module.exports = {
+  async up(queryInterface, Sequelize) {
+    const transaction = await queryInterface.sequelize.transaction();
+    try {
+      await queryInterface.sequelize.query(
+        `DELETE from tree_census WHERE ("plotCensusId","treeId","updatedAt") NOT IN (SELECT "plotCensusId","treeId",MAX("updatedAt") FROM tree_census GROUP BY "plotCensusId","treeId")`
```
Comment (eric-lu-VT) [inline packages/server/src/db/migrations/20220726174811-duplicate-census-constraint.js:8]: maybe add non-SQL explanation of query

File: packages/app/redux/slices/treeCensusSlice.ts [hunk 11]
```diff
@@ -161,6 +161,7 @@ export const upsertTreeCensuses = (
         newState.all[newCensus.id] = newCensus;
         // add to drafts
         if (action?.draft) newState.drafts.add(newCensus.id);
+        else newState.drafts.delete(newCensus.id);
```
Comment (eric-lu-VT) [inline packages/app/redux/slices/treeCensusSlice.ts:164]: Not really sure how the drafts flow work in this file, maybe more comments explaining that

File: packages/app/components/PlottingSheet.tsx [hunk 12]
```diff
@@ -145,61 +223,93 @@ export const PlottingSheet: React.FC<PlottingSheetProps> = ({
...
+      } = isConnected
+        ? await dispatch(createTreeCensus(newCensus))
+        : dispatch(locallyCreateTreeCensus(newCensus));
+      dispatch(selectTree(treeId));
+      dispatch(selectTreeCensus(censusId));
+      expandDrawer();
+    },
+    [isConnected, plotCensus, currentUser, expandDrawer, dispatch, plot.id]
+  );
+  const plotExisting = useCallback(
```
Comment (eric-lu-VT) [inline packages/app/components/PlottingSheet.tsx:265]: rename to plotExistingTree

File: packages/app/components/PlottingSheet.tsx (hunk 6 above)
Comment (eric-lu-VT) [reply packages/app/components/PlottingSheet.tsx:148]: (dir % 4) + 4) % 4

File: packages/server/src/db/migrations/20220810064741-nullable-species-code.js (hunk 9 above)
Comment (julian-george) [reply packages/server/src/db/migrations/20220810064741-nullable-species-code.js:40]: it essentially gives a default value to any null values as the field goes back to being non-nullable

File: packages/server/src/db/migrations/20220726174811-duplicate-census-constraint.js (hunk 10 above)
Comment (julian-george) [reply packages/server/src/db/migrations/20220726174811-duplicate-census-constraint.js:8]: yeah thats a crazy convoluted query haha

File: packages/app/components/FailedModal.tsx (hunk 1 above)
Comment (julian-george) [reply packages/app/components/FailedModal.tsx:62]: will handle in later branch

File: packages/server/src/services/tree-census-service.ts [hunk 13]
```diff
@@ -163,10 +163,6 @@ export const createTreeCensus = async (treeCensus: TreeCensus) => {
 
   const newCensus = await TreeCensusModel.create(treeCensus);
 
-  await Tree.update(
-    { initCensusId: newCensus.id },
-    { where: { id: newCensus.treeId, initCensusId: null } }
-  );
```
Comment (eric-lu-VT) [inline packages/server/src/services/tree-census-service.ts:142]: I need this change committed to master before I can upload data on ReTool

**Review by eric-lu-VT — APPROVED** [review]

**Review by eric-lu-VT — APPROVED** [review]

File: packages/app/components/PlottingSheet.tsx (hunk 5 above)
Comment (julian-george) [reply packages/app/components/PlottingSheet.tsx:141]: will do on different branch

**Pull request merged**